### PR TITLE
perl-bio-ensembl-variation: new package

### PR DESCRIPTION
**Pull request by teaguesterling**
Adding the `perl-bio-ensembl-variation` corresponding to the `Bio::Ensembl::Variation` perl module.

Note about the install approach: this is a Perl module but it doesn't have appropriate installer. In VEP it's downloaded and moved into place by the bundled install.PL script. This is an attempt to make it a bit more conventional.

### CODE REVIEW

**Comment by EbiArnie** [comment]
This should also depend on perl-bio-ensembl.

**Comment by EbiArnie** [comment]
Apart from that, looks good to me.

**Comment by teaguesterling** [comment]
@EbiArnie Thank you for the review! Change made to add `perl-bio-ensembl` as a dependency.

File: var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    version("110", sha256="210d627dcb867d9fda3a0d94428da256f394c32e34df5171b9b9e604507e1f05")
+
+    extends("perl")
+
+    variant("sql", default=False, description="Install SQL files")
+    variant("scripts", default=False, description="Install addition scripts")
+    variant("tools", default=False, description="Install additional tools")
+    variant("ld", default=False, description="Compile LD calculation tools")
+
+    depends_on("perl-bioperl")
```
Comment (EbiArnie) [inline var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py]: The dependencies for these seem to be missing:
perl-bio-bigfile
perl-bio-db-hts
perl-sereal

File: var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    license("APACHE-2.0", checked_by="teaguesterling")
+
+    version("112", sha256="ad75ff0a9efbf2d5c10ab5087d414bac685819664d01fbe4a9765393bd742a7c")
+    version("111", sha256="b2171b3f5f82a2b7e849c0ec8dc254f4bace4b3faba1b3ab75c5eea596e33bef")
+    version("110", sha256="210d627dcb867d9fda3a0d94428da256f394c32e34df5171b9b9e604507e1f05")
+
+    extends("perl")
+
+    variant("sql", default=False, description="Install SQL files")
+    variant("scripts", default=False, description="Install addition scripts")
```
Comment (EbiArnie) [inline var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py]: typo: additional

File: var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        mkdirp(prefix.share.ensembl.variation)
+        for extra in ["sql"]:
+            if spec.satisfies(f"+{extra}"):
+                target = join_path(prefix.share.ensembl, extra)
+                install_tree(extra, target)
+
+        if spec.satisfies("+ld"):
+            mkdirp(prefix.bin)
+            with working_dir("C_code"):
+                install("calc_genotypes", prefix.bin.calc_genotypes)
```
Comment (EbiArnie) [inline var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py]: This doesn't work for me. It seems the 'build' phase is never run. Have you tried installing with "+ld"?


```
==> [2024-07-30-10:39:32.882665] Installing calc_genotypes to /home/arnie/spack/opt/spack/linux-ubuntu22.04-icelake/gcc-11.4.0/perl-bio-ensembl-variation-112-k3iegv4icjabt5bawm5bvsjqz6raxqlc/bin/calc_genotypes
==> Error: OSError: No such file or directory: 'calc_genotypes'

/home/arnie/r/ensembl-swenv-spack-internal/packages/perl-bio-ensembl-variation/package.py:60, in install:
         57        if spec.satisfies("+ld"):
         58            mkdirp(prefix.bin)
         59            with working_dir("C_code"):
  >>     60                install("calc_genotypes", prefix.bin.calc_genotypes)
```

Another thing:
This specifies a subdirectory of prefix.bin with the name calc_genotypes, but that dir does not exist.
Same for the next line.

I think it should be just `prefix.bin`.

**Review by EbiArnie — CHANGES_REQUESTED** [review]
See inline comments - installing with all variants did not work for me.

File: var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py (hunk 3 above)
Comment (teaguesterling) [reply var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py]: Thanks for catching this! I think I refactored this into oblivion and didn't test it properly afterwords. There were two issues that needed to be addressed. The `build` phase needed to be explicitly declared in the `phases` list and `HTSLIB_DIR` environmental variable was missing. Setting those got it to work.

The `install` statement installed the file to the path `prefix.bin.calc_genotypes` as expected, however I think it's not idiomatic. I also realized that I missed handling the `scripts` variant, so likely I'd forgotten this package was a WIP still. Revising and testing to wrap it up.

Comment (teaguesterling) [reply var/spack/repos/builtin/packages/perl-bio-ensembl-variation/package.py]: It *also* turns out that this now depends on `perl-bio-ensembl-io` when running the tools installed with `+tools`. You may have a better/stronger opinion about how this is handled, so suggestions are more than welcome!

**Comment by teaguesterling** [comment]
With this now depending on both `perl-bio-ensembl-io` and `perl-bio-bigfile`, it's probably a good idea to merge this after #44509. I have to merge the dependencies into this to make sure the package audit checks on dependent packages pass as expected.

**Review by EbiArnie — CHANGES_REQUESTED** [review]
Please update this and remove perl-bio-ensembl-io and perl-bio-bigfile.
Please depend on perl-bioperl@1.6.924
Please make sure that each package version depends on the same perl-bio-ensembl and perl-bio-ensembl-io version.

**Comment by teaguesterling** [comment]
> Please update this and remove perl-bio-ensembl-io and perl-bio-bigfile.
> Please depend on perl-bioperl@1.6.924
> Please make sure that each package version depends on the same perl-bio-ensembl and perl-bio-ensembl-io version.

We finally made it to this one. Thank you for summarizing the changes!

**Review by bernhardkaindl — DISMISSED** [review]
@EbiArnie, with the new updates, could you retract your "changes requested" flag that prevents merging by approving this PR?

**Review by bernhardkaindl — COMMENTED** [review]
Built on Ubuntu 22.04 with the update sha

**Review by bernhardkaindl — DISMISSED** [review]
Built on Ubuntu 22.04 with the updated sha.

@EbiArnie, with the new updates, could you retract your "changes requested" flag that prevents merging by approving this PR?

**Comment by EbiArnie** [comment]
I finally found the time to do a proper review and run the test suite.
Apologies for taking so long.

I've found that this depends on two things that are not obvious.
One is Role::Tiny::With, this is not in Spack so far. 
The other is our own "sibling" repo, ensembl-funcgen, perl-bio-ensembl-funcgen in Spack.

Since funcgen is there, it should be added.
For Role::Tiny::With, this could also come later.

In addition, I found quite a few issues in the test suite. I don't think I will have the time to go into all of them.
They have a variety of causes - for example there is code in there for a data pipeline we run.
This code depends on ensembl-hive, but it's absolutely not necessary for anything else.
I think, test failures related to that should be ignored at the moment.

Other errors may be caused by code that's too new, that would most probably be the htslib version in this case.
Most often, the newer version will give better outputs, it's just that they don't match the test suite.

But, I think we can not fix all issues here. So let's go ahead regardless.

Please add the -funcgen dependency.
Then I would go ahead and merge. As future steps, we should add the package for Role::Tiny::With.
And since you care for VEP, I would try a triage issues there so that I don't debug too many things that are not needed for VEP.

**Review by EbiArnie — CHANGES_REQUESTED** [review]
Please add dependency for perl-bio-ensembl-funcgen

**Review by bernhardkaindl — APPROVED** [review]
@EbiArnie Thanks for the review

To get this done, I added as asked for @teaguesterling:
```py
depends_on(f"perl-bio-ensembl-funcgen@{vers}", when=f"@{vers}", type="run")
```

**Comment by teaguesterling** [comment]
> @EbiArnie Thanks for the review
> 
> To get this done, I added as asked for @teaguesterling:
> ```py
> depends_on(f"perl-bio-ensembl-funcgen@{vers}", when=f"@{vers}", type="run")
> ```

Thanks for catching that. I must've missed adding it or forgot to commit.

**Review by EbiArnie — APPROVED** [review]
OK

**Pull request merged**